// saphire/actions/build-registry.js
var mainPackageJSON = require("./main-package.js"),
		mainPackageJSON = mainPackageJSON(),
		path = require("path");

module.exports = function(){
	var saphireRegistry = [],
		i = 0,
		packages = mainPackageJSON["dependencies"];

	var wrapper = [];
	for(saphireCores in packages){
		if(/saphire/g.test(saphireCores)){
			var saphirePackages = saphireCores //resolve and require
			wrapper.push(saphirePackages)
		}
	}

function saphireModulePaths(){
	try {
		//needs to have been init and entry point js file is present
		return require.resolve(wrapper[package])
	} catch (e) {
		return false
	}
}

var saphireRegister = []

for (package in wrapper) {
	var paths = saphireModulePaths();
	var dirs = path.resolve(paths, "../");
	var allPackageJSON = require(dirs+"/package.json")
	saphireRegister[package] = {};
	saphireRegister[package]["name"] = allPackageJSON.name
	saphireRegister[package]["version"] = allPackageJSON.version
	saphireRegister[package]["description"] = allPackageJSON.description
	saphireRegister[package]["repo"] = allPackageJSON.repository
	saphireRegister[package]["saphire"] = allPackageJSON.saphire
}
	return saphireRegister;
}
