// saphire-regions/actions/regions.js
var saphireApi = require("saphire").functions
var saphireData = require("saphire").data

module.exports = function(selectTheme, themeName){

	themeConf = require(selectTheme)
	blocks = require('./blocks.js')


var settingsData = []

var i = 1;
for(prop in region){

	//parse the globals and delete (clear) them if required again and another theme has the same region name.
	//key
	settingsData[prop] = {}

	//name
	settingsData[prop]["name"] = region[prop]
	delete region[prop]

	//Rid
	settingsData[prop]["rid"] = i++

	//classes
	if(classes[prop]){
		settingsData[prop]["class"] = classes[prop]+" region-"+prop
	}else{
		settingsData[prop]["class"] = "region-"+prop
	}
	delete classes[prop] //CLEAR

	//wrapper
	if(wrapper[prop]){
		settingsData[prop]["wrapper"] = wrapper[prop]
	}else{
		settingsData[prop]["wrapper"] = "div"
	}
	delete wrapper[prop] //CLEAR

	//blocks
	if(blocks[prop]){
		settingsData[prop]["blocks"] = blocks[prop]
	}else{
		settingsData[prop]["blocks"] = ""
	}
	delete blocks[prop]

	//data
	if(data[prop]){
		d=0;
		var dataFormatted = []
		for(bit in data[prop]){
			var dataName = Object.keys(data[prop])[d++],
					dataValue = data[prop][bit],
					dataFormat = 'data-'+ dataName +'="'+ dataValue +'"';
					dataFormatted.push(dataFormat)
					settingsData[prop]['dataFormatted'] = dataFormatted.toString().split(",").join(" ");
		}
	}else{
		settingsData[prop]["data"] = ""
	}
	delete data[prop]
}

return settingsData;
}
