// saphire-admin/index.js
saphireRouter = require('./admin-routes/express-short-route.js'),
	express = require('express'),
	saphireMenu = require('saphire').adminMenu,
	util = require('./utils.js'),
	saphireApi = require("saphire").functions;


var saphire = require("saphire"),
		saphireExpose = {
			cores:saphire.cores
		}

var saphireAdmin = {

	entry:function(admin, puroseData){
		//get all data for the admin, this could be pretty much anything
		this.admin = admin;
		this.purposeData = puroseData;
		return this.saphireCores();
	},
	saphireCores:function(){
		var cores = saphireExpose.cores;
		for(core in cores){
			//delete main module info as not needed here
			if(cores[core]["name"] == "saphire"){
				delete cores[core];
			}
		}
		//clean up delete
		var cores = cores.filter(Boolean),
				templateData = []
		//loop clean array
		for(core in cores){
			var friendlyName = cores[core]["name"].replace(/-/g," ");
			var machineName = cores[core]["name"].replace(/-/g,"_");
			var title = cores[core]["name"].split("-")[1];
			var version = cores[core]["version"];
			var description = cores[core]["description"];
			var routerInfo = cores[core]["saphire"]["admin"]["page"];
			//overide this with if query db TO DO
			var menuEnabled = cores[core]["saphire"]["admin"]["menu"]["enabled"];
			var menuIndex = cores[core]["saphire"]["admin"]["menu"]["index"];
			//end overide
			var menu = cores[core]["saphire"]["admin"]["menu"];
			//Build
			templateData[machineName] = {}
			templateData[machineName]["meta"] = {version:version,description:description}
			templateData[machineName]["template"] = {title:title}
			templateData[machineName]["system"] = {machineName:machineName}

			//per purpose pass data to template based on machine name
			var machineName = templateData[machineName]["system"]["machineName"],
					purposeData = this.purposeData[machineName]
			templateData[machineName]["purpose"] = {data:purposeData};

			//get the modules page definitions from the package.json
			templateData[machineName]["routes"] = routerInfo;

			//make admin menu
			saphireApi.extend("adminMenu",machineName,menu)

			//all routes admin common data like menu and such
			templateData[machineName]["common"] = {}
			templateData[machineName]["common"]["menu"] = saphire.data.adminMenu;

		}
		//console.log(this.purposeData)
		this.templateData = templateData;
		return templateData;
	}

}

module.exports = saphireAdmin;
